src/crimson/common/local_shared_foreign_ptr.h

```diff
@@ -0,0 +1,137 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include <seastar/core/smp.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/sharded.hh>
+
+namespace crimson {
+
+/**
+ * local_shared_foreign_ptr
+ *
+ * See seastar/include/seastar/core/sharded.hh:foreign_ptr
+ *
+ * seastar::foreign_ptr wraps a smart ptr by proxying the copy() and destructor
+ * operations back to the original core.  This works well except that copy()
+ * requires a cross-core call.  We need a smart_ptr which allows cross-core
+ * caching of (for example) OSDMaps, but we want to avoid the overhead inherent
+ * in incrementing the source smart_ptr on every copy.  Thus,
+ * local_shared_foreign_ptr maintains a core-local foreign_ptr back to the
+ * original core instance with core-local ref counting.
+ */
+template <typename PtrType>
+class local_shared_foreign_ptr {
```

Comment:
Main problem is that seastar::lw_shared_ptr<seastar::foreign_ptr<T>>::operator* and operator-> seem to return type seastar::foreign_ptr<T>& and seastar::foreign_ptr<T>* rather than T& and T*.  We don't want to have to modify existing users of OSDMapRef, particularly in PeeringState.h/cc.

Reply:
Yeah, repushing with that swapped.